### ISSUE TITLE
Array#pop does not play nice with garbage collection

### ISSUE THREAD

**Issue by ssvb** [manastech/crystal#580]
```ruby
$collected_foo = 0

class Foo
  def finalize
    $collected_foo += 1
  end
end

a = [] of (Foo | Nil)

1000.times { a.push(Foo.new) } # Store a bunch of 'Foo' objects
1000.times { a.pop }           # Remove all of them

GC.collect
puts $collected_foo            #=> 0

1000.times { a.push(nil) }     # Overwrite stale 'Foo' references

GC.collect
puts $collected_foo            #=> 999
```
The "capacity - length" elements in the unused part of the array buffer may still contain leftover junk references to data and prevent this data from being garbage collected.

**Comment by datanoise** [comment]
I believe this is more generate issue. Any array methods that delete elements (shift, delete_if, etc) that modify array's length do no clear the trailing elements which makes them visible to GC.

Also it would be nice to reevaluate the array's capacity when deleting a lot of elements from the array. Otherwise the long living arrays that grow and shirk significantly might occupy a lot of space.

**Comment by Exilor** [comment]
@datanoise shift doesn't appear to have that problem.

**Comment by datanoise** [comment]
well, shift still keeps two references to the last array's element, that could potentially lead to a problem.

**Comment by asterite** [comment]
@ssvb Indeed. We knew this for some time but didn't fix it yet. I thought we would need some primitive to clear a pointer's value to zero (because, what's zero for a particular type?) but I think with sizeof and casting to UInt8 and setting that to zero will work. Probably we'll have `Pointer#clear` and `Pointer#clear(count)` methods. I'll let you know when I implement this. Thanks!

**Comment by asterite** [comment]
I basically did what I said in the last description.

However, memory is set to zero even for primitive types like Int32, which is kind of useless. We have to think of a way to test wether a type T is a reference type or not, but union types make it not-that-trivial to implement (that is, in an elegant way).

**Comment by ssvb** [comment]
@asterite You are really fast. Thanks for the fix!

Just one thing does not feel right. The Crystal array implementation has an extra overhead to explicitly clear the parts of the buffer, which are taken out of use. And at the same time, the garbage collector has an extra overhead to scan the unused buffer areas for pointers. That's a pure waste of CPU cycles on both ends. In a perfect world, the garbage collector preferably needs to be flexible enough to inspect the ```@length``` array property and take it into consideration when scanning for pointers. The [garbage collector API](http://www.hboehm.info/gc/gcinterface.html) provides some support for describing type bitmaps via the ```GC_make_descriptor()``` and ```GC_malloc_explicitly_typed()``` functions, which may be useful for the Crystal structs and tuples, but I can't see anything directly applicable to describe a dynamically resizeable area inside of the allocated buffer.

And you are surely right about the handling of primitive types like Int32. Not doing this extra clearing for the retired array elements and also allocating the buffer via ```GC_malloc_atomic()``` in this case should be a useful performance improvement. Does it make sense to have a new issue filed about this?

**Comment by asterite** [comment]
@ssvb Yes, it makes sense.

However, performance right now is quite good, so these optimizations won't happen soon, unless someone is willing to implement them. Right now our priority is stabilizing things, fixing bugs, completing APIs, documenting stuff and we'll probably need to continue with the dependency manager.

Once all of the above is done, making small optimizations, or replacing the GC with another one (maybe a custom one for this language) should be doable and might bring some performance benefits.
